rpm/0012-Implement-template-addition-for-replacing-keys-in-pr.patch

```diff
@@ -0,0 +1,697 @@
++const char *internal_keys[] = { "HOME", "CFG", "RUNUSER", "PATH", "PRIVILEGED",
++				NULL };
++
++/* Check if the key is in internal key list or it has a hardcoded macro. */
++static int is_internal_macro(const char *key)
++{
++	char *macro;
++	int i;
++
++	for (i = 0; internal_keys[i]; i++) {
++		if (!strcmp(key, internal_keys[i]))
++			return 1;
++	}
++
++	if (asprintf(&macro, "${%s}", key) == -1)
++		errExit("asprintf");
++
++	i = macro_id(macro);
++	free(macro);
++
++	if (i != -1)
++		return 1;
++
++	return 0;
++}
++
++/*
++ *Check the Template argument to have KEY:VALUE in valid format. A valid
++ * Template is added to template list. In case of invalid key, value, internal
++ * macro or existing key firejail is called to exit
++ */
++void check_template(char *arg) {
++	Template *tmpl;
++	const char *key;
++	const char *value;
++	const char *delim = TEMPLATE_KEY_VALUE_DELIM;
++	char *saveptr;
++	int err;
++
++	/* Only alphanumeric chars in template key. */
++	key = strtok_r(arg, delim, &saveptr);
++	if (!is_valid_str(key, STR_CHECK_ALNUM)) {
++		fprintf(stderr, "Error invalid template key \"%s\"\n", key);
++		exit(1);
++	}
++
++	/* Only a-zA-Z0-9_ /*/
++	value = strtok_r(NULL, delim, &saveptr);
++	if (!is_valid_str(value, STR_CHECK_DBUS_COMPAT)) {
++		fprintf(stderr, "Error invalid template value in \"%s:%s\"\n",
++								key, value);
++		exit(1);
++	}
++
++	/* Hardcoded macro or XDG value is not allowed to be overridden. */
++	if (is_internal_macro(key)) {
++		fprintf(stderr, "Error override of \"${%s}\" is not permitted\n",
++									key);
++		exit(1);
++	}
++
++	/* Returns either a Template or exits firejail */
++	tmpl = template_new(key, value);
++
++	err = template_add(tmpl);
++	switch (err) {
++	case 0:
++		return;
++	case -EINVAL:
++		fprintf(stderr, "Error invalid template key \"%s\" "
++						"value \"%s\"\n", key, value);
++		break;
++	case -EALREADY:
++		fprintf(stderr, "Error template key \"%s\" already exists\n",
++								key);
++		break;
++	}
++
++	template_free(tmpl);
++	exit(1);
++}
++
++/*
++ * Check if the argument contains template keys that should be expanded. Will
++ * return 1 only when there is at least one template key found. If an unknown
++ * template exists -EINVAL is returned.  If there is no '$' or the macros are
++ * internal only 0 is returned as there is nothing to expand.
++ */
++int template_requires_expansion(char *arg)
++{
++	char *ptr;
++
++	if (!arg || !*arg)
++		return 0;
++
++	ptr = strchr(arg, '$');
++	if (!ptr)
++		return 0;
++
++	while (*ptr) {
++		if (*ptr == '$' && *(ptr+1) == '{') {
++			char buf[DBUS_MAX_NAME_LENGTH] = { 0 };
++			int err;
++			int i;
++
++			// Copy template key name only
++			for (i = 0, ptr += 2; *ptr && *ptr != '}' &&
++						i < DBUS_MAX_NAME_LENGTH;
++						ptr++, i++)
++				buf[i] = *ptr;
++
++			if (is_internal_macro(buf))
++				continue;
++
++			// Invalid line if '${}' used but no valid template key
++			if (!template_get(buf, &err))
```

Comment:
Here the error is ignored because the line is invalid if there is no internal macro (checked few lines above) and there is no matching template key found. Maybe errno should be utilized instead as in some places that apparently is used.

Reply:
And the main thing is to separate a invalid template key (`EINVAL`)  and not found (`ENOKEY`) from each others. It is used to inform the user about it as a proper error message. `template_requires_expansion()` stops at the first found template key and the line can contain more of those.

Reply:
Modified the patch to use `errno`, how does it look now?

Reply:
Probably okay. The comments need updating as now they still talk about `err` and the topmost functions don't mention `errno` at all although that's how they communicate the type of error.

Reply:
Good that you spotted those, completely forgot them. Changed the comments on the functions.

Reply:
Updated the comments more and added also man page descriptions of this as well, which are required when attempting to upstream this. No harm having them there even though we do not use them. Still, we can locally utilize the man pages if needed.